### PR TITLE
swa object detection

### PR DESCRIPTION
**Pull request by hyz-xmaster**
Recently, we found the simple technique, [Stochastic Weights Averaging](https://arxiv.org/abs/1803.05407) (SWA),  was very effective in object detection. Through extensive experiments, we discovered a good policy of applying it to object detection, which can consistently achieve ~1.0 AP improvement over various object detectors on the COCO benchmark.  The policy is surprisingly simple: **training your detector for an extra 12 epochs using cyclical learning rates and then averaging these 12 checkpoints as your final detection model**. This does not incur any inference cost and is really worth trying. To find full details, please have a look at our report at [https://arxiv.org/abs/2012.12645](https://arxiv.org/abs/2012.12645).

Therefore, we implemented the SWA training with MMDetection so that the community (#2001) can benefit from this simple technique. The idea of the implementation can be found [here](https://github.com/hyz-xmaster/swa_object_detection#instructions). 

I create this pull request also as the discussion thread, according to the Contributing Guideline. If you think this PR is not suitable, I can close it then.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @hyz-xmaster ,
Thanks for your kind PR. We will review it ASAP.

**Comment by Johnson-Wang** [comment]
Hi thanks for your contribution. This work is impressive. There are a few comments after an overview of this PR. 
- I note that the improvement of 1x training + 1x finetuning with swa would increase the performance by about 1% AP. However, the performance of a direct 2x training is actually comparable to that. I wonder if it is possible to show the comparison given the same training budget. For example, the 1x training + 1x swa finetuning should be compared to direct 2x training, and 2x training + 1x swa finetuning should be compared to direct 3x training. 
- There are some changes in `mmdet/apis/train.py`. This could cause multiple compatibility issues for other algorithms and downstream repos. I suggest reformatting SWAHook or adding a new feature in [mmcv runners](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/epoch_based_runner.py).

**Comment by hyz-xmaster** [comment]
> Hi thanks for your contribution. This work is impressive. There are a few comments after an overview of this PR.
> 
> * I note that the improvement of 1x training + 1x finetuning with swa would increase the performance by about 1% AP. However, the performance of a direct 2x training is actually comparable to that. I wonder if it is possible to show the comparison given the same training budget. For example, the 1x training + 1x swa finetuning should be compared to direct 2x training, and 2x training + 1x swa finetuning should be compared to direct 3x training.
> * There are some changes in `mmdet/apis/train.py`. This could cause multiple compatibility issues for other algorithms and downstream repos. I suggest reformatting SWAHook or adding a new feature in [mmcv runners](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/epoch_based_runner.py).

Hi, thanks for your comments. 

* Good point. In fact, we have considered this, so we use the Mask RCNN-2x instead of the Mask RCNN-1x as our baseline detector for ablation experiments in our report. If you have a close look at our report, you can find the implicit comparison of 1x training + 1x swa with 2x training, and the former performs better than the latter. Even though we do not conduct the comparison between 2x training + 1x swa and 3x training, which is time-consuming, we believe the former will still have a better performance than the latter. 
* For the implementation, I have thought a lot about the flexibility and  the amount of changes to the codebase, and the current implementation is what I can think of. MMCV runner is a more basic function so I did not work on it. For the reimplementation, I hope you can help with it, because I am not a professional coder and do not have a clear overview of the whole code frame. In contrast, you pros may have much better ideas of the implementation. So look forward to your shot.😀

**Comment by Johnson-Wang** [comment]
The [ema hook](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/ema.py) may help you to restructure your code. We may target on wrapping all pipelines in a single hook class.
